### PR TITLE
Messages

### PR DESCRIPTION
**Pull request by Taylor-Girard**

### CODE REVIEW

File: app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java [hunk 1]
```diff
@@ -1,17 +1,213 @@
...
+        queryReceiver.whereEqualTo("Sender", match);
+        queryReceiver.whereEqualTo("Receiver", user);
+        queryList.add(queryReceiver);
+
+        ParseQuery<Message> finalQuery = ParseQuery.or(queryList);
+
+        SubscriptionHandling<Message> subscriptionHandling = parseLiveQueryClient.subscribe(finalQuery);
+
+        subscriptionHandling.handleEvent(SubscriptionHandling.Event.CREATE, (query, object) -> {
+            messages.add(0, object);
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java:180]: do we handle moving existing messages being updated?

File: app/src/main/java/com/taylorgirard/comicconvo/activities/MainActivity.java [hunk 2]
```diff
@@ -31,7 +32,7 @@ protected void onCreate(Bundle savedInstanceState) {
         setContentView(R.layout.activity_main);
 
         final Fragment[] fragment = new Fragment[1];
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/activities/MainActivity.java:34]: why is this a list?

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+        if (viewType == MESSAGE_INCOMING){
+            View contactView = inflater.inflate(R.layout.message_incoming, parent, false);
+            return new IncomingMessageViewHolder(contactView);
+        } else if (viewType == MESSAGE_OUTGOING){
+            View contactView = inflater.inflate(R.layout.message_outgoing, parent, false);
+            return new OutgoingMessageViewHolder(contactView);
+        } else{
+            throw new IllegalArgumentException("Unknown view type");
+        }
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java:58]: switch case

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    @Override
+    public int getItemViewType(int position) {
+        if (isMe(position)){
+            return MESSAGE_OUTGOING;
+        } else {
+            return MESSAGE_INCOMING;
+        }
+    }
+
+    private boolean isMe(int position){
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java:76]: how does this look compared to sending to users that are not yourself?

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                name.setText(message.getSender().fetchIfNeeded().getUsername());
+            } catch (ParseException e) {
+                e.printStackTrace();
+            }
+            ParseFile profilePic = null;
+            try {
+                profilePic = message.getSender().fetchIfNeeded().getParseFile("profilePic");
+            } catch (ParseException e) {
+                e.printStackTrace();
+            }
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java:116]: can we put this all in one try catch?

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            if (profilePic != null) {
+                Glide.with(context).load(profilePic.getUrl()).transform(new CircleCrop()).into(imageOther);
+            }
+        }
+    }
+
+    public class OutgoingMessageViewHolder extends MessageViewHolder{
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java:123]: Let's move `OutgoingMessageViewHolder` and `IncomingMessageViewHolder` into their own files

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            return MESSAGE_INCOMING;
+        }
+    }
+
+    private boolean isMe(int position){
+        Message message = messages.get(position);
+        return message.getSenderId() != null && message.getSenderId().equals(user.getObjectId());
+    }
+
+    public abstract class MessageViewHolder extends RecyclerView.ViewHolder{
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java:81]: Let's move this into its own file

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                }
+                //Create intent to go to the individual message activity
+                Intent intent = new Intent(context, IndividualMessageActivity.class);
+                intent.putExtra("Match", match);
+                context.startActivity(intent);
+
+            }
+        }
+
+        public boolean isUser(Message message){
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java:111]: Could we use a more descriptive name?

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            String senderId = message.getSenderId();
+            String userId = ParseUser.getCurrentUser().getObjectId();
+
+            if (!senderId.equals(userId)){
+                return false;
+            } else {
+                return true;
+            }
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java:119]: I think we can shorten this to
`return senderId.equals(userId)`

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        public void bind(Message message){
+            ParseUser match;
+            Boolean isUser = isUser(message);
+            if (!isUser){
+                match = message.getSender();
+            } else {
+                match = message.getReceiver();
+            }
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java:75]: You can use ternary operators for short if-else statements: 
``` 
match = (isUser) ? message.getReceiver() : message.getSender(); 
```

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            try {
+                tvMatchName.setText(match.fetchIfNeeded().getUsername());
+            } catch (ParseException e) {
+                e.printStackTrace();
+            }
+            if (isUser){
+                tvMessageBody.setText("You: " + message.getBody());
+            } else{
+                tvMessageBody.setText(match.getUsername() + ": " + message.getBody());
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java:84]:
```
tvMessageBody.setText(
     (isUser) ? "You: " + message.getBody() : 
     match.getUsername() + ": " + message.getBody());
```

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                Message message = messages.get(position);
+                Boolean isUser = isUser(message);
+                if (!isUser){
+                    match = message.getSender();
+                } else {
+                    match = message.getReceiver();
+                }
+                //Create intent to go to the individual message activity
+                Intent intent = new Intent(context, IndividualMessageActivity.class);
+                intent.putExtra("Match", match);
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageListAdapter.java:105]: could we move all extra names ("Match", etc.) into a helper file?

File: app/src/main/java/com/taylorgirard/comicconvo/fragments/MessagesFragment.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        rvMessageList.setAdapter(adapter);
+
+        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
+        //linearLayoutManager.setReverseLayout(true);
+        rvMessageList.setLayoutManager(linearLayoutManager);
+    }
+
+    public void makeMessageList(){
+
+        List<ParseUser> messaged = user.getList("Messaged");
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/fragments/MessagesFragment.java:62]: is this nullable?

File: app/src/main/java/com/taylorgirard/comicconvo/fragments/MessagesFragment.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        super.onViewCreated(view, savedInstanceState);
+
+        rvMessageList = view.findViewById(R.id.rvMessageList);
+        makeMessageList();
+
+        adapter = new MessageListAdapter(getContext(), recents);
+        rvMessageList.setAdapter(adapter);
+
+        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
+        //linearLayoutManager.setReverseLayout(true);
```
Comment (shenwang00) [inline app/src/main/java/com/taylorgirard/comicconvo/fragments/MessagesFragment.java:56]: remove unused code

File: app/src/main/res/drawable/message_outgoing.xml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <solid android:color="@color/design_default_color_primary" />
+    <corners android:topRightRadius="5dp" android:radius="15dp" />
```
Comment (shenwang00) [inline app/src/main/res/drawable/message_outgoing.xml:5]: can the corner dimensions be moved into `dimens.xml` and shared by `message_incoming` and `message_outgoing`?

File: app/src/main/res/layout/activity_individual_message.xml [hunk 16]
```diff
@@ -6,10 +6,57 @@
...
-        android:layout_width="wrap_content"
+    <RelativeLayout
+        android:id="@+id/rlSend"
+        android:layout_width="match_parent"
         android:layout_height="wrap_content"
-        android:text="Empty message activity"
+        android:layout_alignParentBottom="true"
+        android:background="#FFFFFF"
+        android:elevation="@dimen/body_elevation"
+        android:paddingLeft="0dp"
```
Comment (shenwang00) [inline app/src/main/res/layout/activity_individual_message.xml:16]: do we need to add padding left and right if they're 0?

File: app/src/main/res/layout/item_message.xml [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+    <TextView
+        android:id="@+id/tvMessageBody"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/body_margin_left"
+        android:layout_marginTop="@dimen/body_padding_vertical"
+        app:layout_constraintStart_toEndOf="@+id/ivMatchPhoto"
+        app:layout_constraintTop_toBottomOf="@+id/tvMatchName"
+        tools:text="This is my text" />
```
Comment (shenwang00) [inline app/src/main/res/layout/item_message.xml:34]: is this needed?

File: app/src/main/res/values/strings.xml [hunk 18]
```diff
@@ -22,9 +23,12 @@
...
     <string name="convention_checkbox">Convention</string>
     <string name="meetup_checkbox">Meetup</string>
+    <string name="message_hint">Send a message</string>
     <string-array name="tag_list">
         <item name="store_tag">Store</item>
         <item name="convention_tag">Convention</item>
         <item name="meetup_tag">Meetup</item>
     </string-array>
+    <!-- TODO: Remove or change this placeholder text -->
+    <string name="hello_blank_fragment">Hello blank fragment</string>
```
Comment (shenwang00) [inline app/src/main/res/values/strings.xml:33]: can be removed

File: app/src/main/java/com/taylorgirard/comicconvo/fragments/MessagesFragment.java (hunk 13 above)
Comment (Taylor-Girard) [reply app/src/main/java/com/taylorgirard/comicconvo/fragments/MessagesFragment.java:62]: "Messaged" has default value of an empty list, so it can only return back an empty list or a full one

File: app/src/main/res/layout/item_message.xml (hunk 17 above)
Comment (Taylor-Girard) [reply app/src/main/res/layout/item_message.xml:34]: This is just used for me to see a default value in the design mode so I can better understand what it will look like in practice

File: app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java [hunk 19]
```diff
@@ -1,17 +1,213 @@
...
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_individual_message);
+        setupMessagePosting();
+        loadMessages();
+    }
+
+    @Override
+    public void onBackPressed() {
+        Intent intent = new Intent(IndividualMessageActivity.this, MainActivity.class);
```
[the remark this answers is not in the document]
Comment (Taylor-Girard) [reply app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java:60]: I don't want the back press functionality to be what it usually is because then the information will not be updated. Instead I want the user to be pushed back to the homepage so that when they reload their message list the new changes are reflected

File: app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java [hunk 20]
```diff
@@ -1,17 +1,213 @@
...
+
+                String userId = user.getObjectId();
+                String matchId = match.getObjectId();
+                String pairId;
+
+                if (userId.compareTo(matchId) < 0){
+                    pairId = userId + matchId;
+                } else {
+                    pairId = matchId + userId;
+                }
```
[the remark this answers is not in the document]
Comment (Taylor-Girard) [reply app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java:107]: I want the pairId to be the same for each pair, so I just combine the ids in sorted order

File: app/src/main/java/com/taylorgirard/comicconvo/activities/MainActivity.java (hunk 2 above)
Comment (Taylor-Girard) [reply app/src/main/java/com/taylorgirard/comicconvo/activities/MainActivity.java:34]: It would not let me edit the fragment within inner classes unless it was "effectively final" and this was the way android studio's quick fix recommended I do so

File: app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java (hunk 1 above)
Comment (Taylor-Girard) [reply app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java:180]: I'm not quite sure what you mean, but messages that have already been sent cannot be edited

File: app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java (hunk 4 above)
Comment (Taylor-Girard) [reply app/src/main/java/com/taylorgirard/comicconvo/adapters/MessageAdapter.java:76]: This simply checks whether the sender of the message is you to pick the correct ViewHolder for the message. The message layout is different depending on if it is outgoing or incoming. You cannot send messages to yourself.

File: app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (Taylor-Girard) [reply app/src/main/java/com/taylorgirard/comicconvo/activities/IndividualMessageActivity.java:107]: A collision may be possible, but given that the user ids are strings of 10 random characters, I think this may be unlikely to occur.
